### PR TITLE
HPCC-32060 Rationalize fastThrough lookahead logic

### PR DESCRIPTION
**Pull request by jakesmith**
Modify info meta flags and logic, to avoid some unnecessary lookaheads

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - for discussion. Please take a look

File: thorlcr/activities/aggregate/thaggregateslave.cpp [hunk 1]
```diff
@@ -196,7 +196,6 @@ class AggregateSlaveActivity : public AggregateSlaveBase
     virtual void getMetaInfo(ThorDataLinkMetaInfo &info) const override
     {
         initMetaInfo(info);
-        info.singleRowOutput = true;
```
Comment (jakesmith) [inline thorlcr/activities/aggregate/thaggregateslave.cpp:199]: removing excessive unused flags (not used anywhere), and can be determined by totalRowsMin==totalRowsMax==1

File: thorlcr/activities/choosesets/thchoosesetsslave.cpp [hunk 2]
```diff
@@ -308,7 +308,6 @@ class ChooseSetsPlusActivity : public CSlaveActivity, implements ILookAheadStopN
     virtual void getMetaInfo(ThorDataLinkMetaInfo &info) const override
     {
         initMetaInfo(info);
-        info.buffersInput = true;
```
Comment (jakesmith) [inline thorlcr/activities/choosesets/thchoosesetsslave.cpp:311]: buffersInput and canBufferInput were used interchangeably, wherever buffersInput OR canBufferInput existed, it was to inform a decision re. canStall()

File: thorlcr/activities/filter/thfilterslave.cpp [hunk 3]
```diff
@@ -157,6 +157,11 @@ class CFilterSlaveActivity : public CFilterSlaveActivityBase, public CThorSteppa
         CThorSteppable::setInputStream(index, input, consumerOrdered);
     }
     virtual IInputSteppingMeta *querySteppingMeta() { return CThorSteppable::inputStepping; }
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) const override
+    {
+        PARENT::getMetaInfo(info);
+        info.fastThrough = true;
```
Comment (jakesmith) [inline thorlcr/activities/filter/thfilterslave.cpp:163]: ok, filter isn't "fast through", but for the way it's being used I think filter and project should be.

File: thorlcr/activities/firstn/thfirstnslave.cpp [hunk 4]
```diff
@@ -124,6 +124,11 @@ class CFirstNSlaveLocal : public CFirstNSlaveBase
         }
         return NULL;
     }
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) const override
+    {
+        PARENT::getMetaInfo(info);
+        info.fastThrough = true;
```
Comment (jakesmith) [inline thorlcr/activities/firstn/thfirstnslave.cpp:130]: NB: this local firstn

File: thorlcr/activities/limit/thlimitslave.cpp [hunk 5]
```diff
@@ -75,7 +75,6 @@ class CLimitSlaveActivityBase : public CSlaveActivity
     {
         initMetaInfo(info);
         info.canReduceNumRows = true;
-        info.canBufferInput = false;
```
Comment (jakesmith) [inline thorlcr/activities/limit/thlimitslave.cpp:78]: default anyway

File: thorlcr/activities/nsplitter/thnsplitterslave.cpp [hunk 6]
```diff
@@ -208,6 +208,8 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
 
                 PARENT::start();
                 initMetaInfo(cachedMetaInfo);
+                cachedMetaInfo.suppressLookAhead = true;
```
Comment (jakesmith) [inline thorlcr/activities/nsplitter/thnsplitterslave.cpp:211]: will suppress ensureFTLookAhead downstream from inserting a lookahead

File: thorlcr/activities/pull/thpullslave.cpp [hunk 7]
```diff
@@ -54,7 +54,9 @@ class PullSlaveActivity : public CSlaveActivity
     virtual void getMetaInfo(ThorDataLinkMetaInfo &info) const override
     {
         initMetaInfo(info);
-        info.buffersInput = true;
+        info.canBufferInput = true;
+        info.suppressLookAhead = true;
```
Comment (jakesmith) [inline thorlcr/activities/pull/thpullslave.cpp:58]: will suppress ensureFTLookAhead downstream from inserting a lookahead

**Review by ghalliday — APPROVED** [review]

**Pull request merged**